File: backend/src/main/java/com/apocalypse/thefall/entity/Tile.java
package com.apocalypse.thefall.entity;

import jakarta.persistence.*;
import lombok.*;

import java.util.Collections;

@Entity
@Getter
@Setter
@ToString(exclude = {"map", "terrainConfiguration"})
@Table(name = "tile")
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Tile {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne
    @JoinColumn(name = "map_id")
    private Map map;

    private int x;
    private int y;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "terrain_configuration_id", nullable = false)
    private TerrainConfiguration terrainConfiguration;

    public boolean isWalkable() {
        return terrainConfiguration.isWalkable();
    }

    public int getMovementCost() {
        return terrainConfiguration.getMovementCost();
    }

    public static Tile unknownAt(int x, int y, Map map) {
        return Tile.builder()
                .x(x)
                .y(y)
                .map(map)
                .terrainConfiguration(TerrainConfiguration.builder()
                        .name("unknown")
                        .walkable(false)
                        .movementCost(999)
                        .descriptions(Collections.emptyMap())
                        .build())
                .build();
    }

}

File: backend/src/main/java/com/apocalypse/thefall/entity/Tile.kt
package com.apocalypse.thefall.entity

import jakarta.persistence.*

@Entity
@Table(name = "tile")
open class Tile {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    open var id: Long? = null

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "map_id")
    open var map: GameMap? = null

    @Column(nullable = false)
    open var x: Int = 0

    @Column(nullable = false)
    open var y: Int = 0

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "terrain_configuration_id", nullable = false)
    open var terrainConfiguration: TerrainConfiguration? = null

    open val movementCost: Int? get() = terrainConfiguration?.movementCost

    open val isWalkable: Boolean? get() = terrainConfiguration?.walkable

    companion object {
        fun unknownAt(x: Int, y: Int, map: GameMap): Tile {
            val terrain = TerrainConfiguration().apply {
                name = "unknown"
                walkable = false
                movementCost = 999
                descriptions = mutableMapOf()
                path = ""
            }
            return Tile().apply {
                this.x = x
                this.y = y
                this.map = map
                this.terrainConfiguration = terrain
            }
        }
    }
}
